File: java/3412.java
class Solution {
    public long calculateScore(String s) {
        int n = s.length();
        int stack[][] = new int[26][n];
        int ind[] = new int[26];
        long sum = 0;
        for (int i = 0; i < n; i++) {
            int t1 = s.charAt(i) - 97;
            int t2 = 25 - t1;
            if (ind[t2] == 0)
                stack[t1][ind[t1]++] = i;
            else
                sum += i - stack[t2][--ind[t2]];
        }
        return sum;
    }
}

File: cpp/3412.cpp
class Solution {
public:
    long long calculateScore(string& s) {
        int n = s.length(), t;
        int* stack[26];
        for (int i = 0; i < 26; i++)
            stack[i] = (int*)malloc(sizeof(int) * n);
        int ind[26] = {0};
        long long sum = 0;
        for (int i = 0; i < n; i++) {
            t = s[i] - 97;
            if (ind[25 - t] == 0)
                stack[t][ind[t]++] = i;
            else
                sum += i - stack[25 - t][--ind[25 - t]];
        }
        return sum;
    }
};
